python-checks/src/main/java/org/sonar/python/checks/tests/AssertAfterRaiseCheck.java

```diff
@@ -0,0 +1,101 @@
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.python.checks.tests;
+
+import java.util.List;
+import java.util.Optional;
+import org.sonar.check.Rule;
+import org.sonar.plugins.python.api.IssueLocation;
+import org.sonar.plugins.python.api.PythonSubscriptionCheck;
+import org.sonar.plugins.python.api.tree.CallExpression;
+import org.sonar.plugins.python.api.tree.ExpressionStatement;
+import org.sonar.plugins.python.api.tree.Name;
+import org.sonar.plugins.python.api.tree.QualifiedExpression;
+import org.sonar.plugins.python.api.tree.Statement;
+import org.sonar.plugins.python.api.tree.Tree;
+import org.sonar.plugins.python.api.tree.WithStatement;
+import org.sonar.python.tests.PytestUtils;
+import org.sonar.python.tests.UnittestUtils;
+
+@Rule(key = "S5915")
+public class AssertAfterRaiseCheck extends PythonSubscriptionCheck {
+  private static final String MESSAGE_MULTIPLE_STATEMENT = "Don’t perform an assertion here; An exception is expected to be raised before its execution.";
+  private static final String MESSAGE_SINGLE_STATEMENT = "Refactor this test; if this assertion’s argument raises an exception, the assertion will never get executed.";
+  private static final String MESSAGE_SECONDARY = null;
+
+  @Override
+  public void initialize(Context context) {
+    context.registerSyntaxNodeConsumer(Tree.Kind.WITH_STMT, ctx -> {
+      WithStatement withStatement = (WithStatement) ctx.syntaxNode();
+      if (!isWithStatementItemARaise(withStatement)) {
+        return;
+      }
+
+      List<Statement> statements = withStatement.statements().statements();
+      Statement statement = statements.get(statements.size()-1);
+      if (isAnAssert(statement)) {
+        ctx.addIssue(statement, statements.size() > 1 ? MESSAGE_MULTIPLE_STATEMENT : MESSAGE_SINGLE_STATEMENT)
+          .secondary(IssueLocation.preciseLocation(withStatement.firstToken(), withStatement.colon(), MESSAGE_SECONDARY));
+      }
+    });
+  }
+
+  public boolean isWithStatementItemARaise(WithStatement withStatement) {
+    return withStatement.withItems().stream()
+      .filter(withItem -> withItem.test().is(Tree.Kind.CALL_EXPR))
+      .map(withItem -> ((CallExpression) withItem.test()).callee())
+      .filter(callee -> callee.is(Tree.Kind.QUALIFIED_EXPR))
+      .map(QualifiedExpression.class::cast)
+      .anyMatch(callee -> isPytestRaise(callee) || isUnittestRaise(callee));
+  }
+
+  public boolean isPytestRaise(QualifiedExpression callee) {
+    return PytestUtils.isPytest(callee) && PytestUtils.RAISE_METHODS.contains(callee.name().name());
```

Comment:
I think just checking that the callee has `pytest.raises` as fully qualified name is enough. We can also just test that `pytest.raises` is in the FQN without being strictly equal, to account for potential wrappers.
In any case, we probably don't expect a lot of false positive here: this pattern seems to be a mistake for any framework.